src/envoy/auth/http_filter.cc

```diff
@@ -53,34 +53,29 @@ FilterHeadersStatus JwtVerificationFilter::decodeHeaders(HeaderMap& headers,
   state_ = Calling;
   stopped_ = false;
 
-  /*
-   * TODO: update cached public key regularly
-   * TODO: handle the case that the next request comes before fetching public
-   * key completed
-   */
-
-  // list up issuers whose public key should be fetched
+  // list up issuers whose public key should be fetched.
+  bool at_least_one_call = false;
+  all_issuers_pubkey_expiration_checked_ = false;
+  uint32_t index = 0;
   for (const auto& iss : config_->issuers_) {
-    if (!iss->failed_ && !iss->loaded_) {
-      calling_issuers_[iss->name_] = std::make_pair(iss, nullptr);
+    if (++index == config_->issuers_.size()) {
+      all_issuers_pubkey_expiration_checked_ = true;
     }
-  }
-  // send HTTP requests to fetch public keys
-  if (!calling_issuers_.empty()) {
-    for (const auto& iss : config_->issuers_) {
-      if (iss->failed_ || iss->loaded_) {
-        continue;
-      }
-      calling_issuers_[iss->name_].second =
-          std::unique_ptr<Auth::AsyncClientCallbacks>(
-              new Auth::AsyncClientCallbacks(
-                  config_->cm_, iss->cluster_,
-                  [&](bool succeed, const std::string& pubkey) -> void {
-                    this->ReceivePubkey(headers, iss->name_, succeed, pubkey);
-                  }));
-      calling_issuers_[iss->name_].second->Call(iss->uri_);
+    if (!iss->pkey_->IsNotExpired()) {
```

Comment:
This will be easily deadlocked when request A comes and the filter issued requests to public key server, then another request with same JWT comes and reached here, the thread is locked and never being unblocked since you won't be able to process the response from public key server.